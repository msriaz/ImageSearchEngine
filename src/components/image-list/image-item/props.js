import PropTypes from 'prop-types';

export const propTypes = {
  imageData: PropTypes.object.isRequired,
};

export const defaultProps = {};
