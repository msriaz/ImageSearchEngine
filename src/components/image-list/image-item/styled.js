import {Spinner} from '../../spinner';
import styled from 'styled-components/native';
import FastImage from 'react-native-fast-image';

export const Image = styled(FastImage)`
  width: 45%;
  height: ${({theme}) => theme.windowWidth * 0.45}px;
  resize-mode: cover;
  margin: ${({theme}) => theme.spacing.small}px;
`;

export const Loader = styled(Spinner)`
  left: 50%;
  margin-left: -${({theme}) => theme.spacing.small}px;
  margin-top: -${({theme}) => theme.spacing.small}px;
  position: absolute;
  top: 50%;
  z-index: 2;
  margin: ${({theme}) => theme.spacing.small}px;
`;

export const Wrapper = styled.View`
  height: ${({theme}) => theme.windowWidth * 0.45}px;
  position: absolute;
  width: 45%;
`;

export const PreviewImage = styled.View`
  align-items: center;
  background-color: ${({theme}) => theme.color.lightestGray};
  height: ${({theme}) => theme.windowWidth * 0.45}px;
  justify-content: center;
  position: absolute;
  width: 45%;
  margin: ${({theme}) => theme.spacing.small}px;
`;
