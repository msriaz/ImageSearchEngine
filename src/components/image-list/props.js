import PropTypes from 'prop-types';

export const propTypes = {
  data: PropTypes.array.isRequired,
  loadMore: PropTypes.func.isRequired,
  isLoadingMore: PropTypes.bool.isRequired,
};

export const defaultProps = {};
