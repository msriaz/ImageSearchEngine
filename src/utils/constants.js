import {color as colors} from '../theme';
import {generateColorPalleteConstants} from './index';

export const APP_COLORS = generateColorPalleteConstants(colors);

export const STORAGE_KEYS = {
  HISTORY: 'history',
};
