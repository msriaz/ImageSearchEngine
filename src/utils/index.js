import {useEffect, useRef} from 'react';
import snakeCase from 'lodash/snakeCase';
/**
 *
 * @param {*} value
 * @returns previous state
 */
export const usePrevious = value => {
  const ref = useRef();
  useEffect(() => {
    ref.current = value;
  });
  return ref.current;
};

/**
 *
 * @param {*} imgData
 * @returns image url to render
 */
export const getImageUrl = imgData => {
  const {farm, server, id, secret} = imgData;
  return `https://farm${farm}.static.flickr.com/${server}/${id}_${secret}.jpg`;
};

/**
 * generate color consts from color pallette
 * @param {*} colors
 */
export const generateColorPalleteConstants = colors => {
  const colorsList = {};
  Object.keys(colors).forEach(colorKey => {
    colorsList[snakeCase(colorKey).toUpperCase()] = colors[colorKey];
  });
  return colorsList;
};

/**
 *
 * @param {*} funcName
 * @param {*} error
 */
export const reportError = (funcName, error) => {
  console.log(`Error in ${funcName}`, error);
};
