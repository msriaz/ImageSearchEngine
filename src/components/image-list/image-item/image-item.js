import React, {useState} from 'react';
import {APP_COLORS} from '../../../utils/constants';

import {propTypes, defaultProps} from './props';
import {Image, Loader, PreviewImage, Wrapper} from './styled';
import {getImageUrl} from '../../../utils';

/**
 * Image Item Component
 * @param {imageData} defines the data of image to render
 */
export const ImageItem = ({imageData}) => {
  const [isLoading, setIsLoading] = useState(true);

  const path = getImageUrl(imageData);

  return (
    <React.Fragment>
      {isLoading && (
        <Wrapper>
          <Loader color={APP_COLORS.GRAY} size="small" />
          <PreviewImage>
            <Image style={{position: 'absolute'}} source={{uri: path}} />
          </PreviewImage>
        </Wrapper>
      )}
      <Image
        source={{uri: path}}
        onLoadStart={() => setIsLoading(true)}
        onLoad={() => setIsLoading(false)}
      />
    </React.Fragment>
  );
};

ImageItem.propTypes = propTypes;
ImageItem.defaultProps = defaultProps;
