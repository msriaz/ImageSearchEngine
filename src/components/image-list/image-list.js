import React, {forwardRef} from 'react';
import {FlatList} from 'react-native';
import {props, defaultProps} from './props';

import {ImageItem} from './image-item';
import {Spinner} from '../spinner';

/**
 * Image Item Component
 * @param {data} defines the data of images to render
 * @param {loadMore} defines the function to be called for loading more data
 * @param {rest} defines additional props to pass to FlatList
 * @param {isLoadingMore} defines to load more
 */
export const ImageList = forwardRef(
  ({data, loadMore, isLoadingMore, rest}, ref) => {
    const renderItem = ({item}) => <ImageItem imageData={item} key={item.id} />;

    const ListFooterComponent = () =>
      isLoadingMore ? <Spinner variant={Spinner.VARIANTS.SMALL} /> : null;

    return (
      <FlatList
        ref={ref}
        data={data}
        renderItem={renderItem}
        horizontal={false}
        numColumns={2}
        onEndReached={loadMore}
        ListFooterComponent={ListFooterComponent}
        {...rest}
      />
    );
  },
);

ImageList.props = props;
ImageList.defaultProps = defaultProps;
