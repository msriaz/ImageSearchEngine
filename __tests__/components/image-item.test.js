import React from 'react';
import {ImageItem} from '../../src/components/image-list/image-item';
import {render} from '../utils/theme-wrapper';

const mockData = {
  id: '51443585974',
  secret: '0e481ab487',
  server: '65535',
  farm: 66,
};

describe('Image Item Component', () => {
  it('should match the Image Item  snapshot', () => {
    const compTree = render(<ImageItem imageData={mockData} />);
    expect(compTree.toJSON()).toMatchSnapshot();
  });
});
