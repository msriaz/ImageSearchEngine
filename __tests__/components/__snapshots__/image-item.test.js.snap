// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image Item Component should match the Image Item  snapshot 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "height": "NaNpx",
          "position": "absolute",
          "width": "45%",
        },
      ]
    }
  >
    <ActivityIndicator
      color="#A8A8A8"
      size="small"
      style={
        Array [
          Object {
            "left": "50%",
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
            "position": "absolute",
            "top": "50%",
            "zIndex": 2,
          },
        ]
      }
    />
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#F4F4F4",
            "height": "NaNpx",
            "justifyContent": "center",
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
            "position": "absolute",
            "width": "45%",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Array [
              Object {
                "height": "NaNpx",
                "marginBottom": 8,
                "marginLeft": 8,
                "marginRight": 8,
                "marginTop": 8,
                "resizeMode": "cover",
                "width": "45%",
              },
              Object {
                "position": "absolute",
              },
            ],
          ]
        }
      >
        <FastImageView
          resizeMode="cover"
          source={
            Object {
              "uri": "https://farm66.static.flickr.com/65535/51443585974_0e481ab487.jpg",
            }
          }
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        />
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Array [
          Object {
            "height": "NaNpx",
            "marginBottom": 8,
            "marginLeft": 8,
            "marginRight": 8,
            "marginTop": 8,
            "resizeMode": "cover",
            "width": "45%",
          },
        ],
      ]
    }
  >
    <FastImageView
      onFastImageLoad={[Function]}
      onFastImageLoadStart={[Function]}
      resizeMode="cover"
      source={
        Object {
          "uri": "https://farm66.static.flickr.com/65535/51443585974_0e481ab487.jpg",
        }
      }
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    />
  </View>,
]
`;
